File: ontology-data-access-suite-full-v3.3-refactored-v2/ontology-data-access-suite-full-v3.3-refactored/shared/oql_validator.py

#!/usr/bin/env python3
"""Structural validator for canonical OQL (expression-aware)."""
from __future__ import annotations
import argparse, json
from pathlib import Path
from typing import Any, Dict, List, Set
READ_OPS = {"QUERY", "AGGREGATE", "ASSOCIATION_QUERY", "LINK_QUERY"}
WRITE_OPS = {"CREATE", "UPDATE", "DELETE", "UPSERT", "BATCH"}
ALL_OPS = READ_OPS | WRITE_OPS
ALLOWED_TOP = {"version","schemaRef","strict","operation","objects","relationships","conditions","returns","orders","maxResults","sourceQuery","linkQuery","mutation","options","extensions"}
CONDITION_OPERATORS = {"EQ":1,"NE":1,"GT":1,"GTE":1,"LT":1,"LTE":1,"IN":"many","NOT_IN":"many","BETWEEN":2,"LIKE":1,"CONTAINS":1,"STARTS_WITH":1,"ENDS_WITH":1,"IS_NULL":0,"IS_NOT_NULL":0}
RETURN_KINDS = {"FIELDS","EXPR","GROUP_BY","METRIC"}
METRIC_FUNCTIONS = {"COUNT","SUM","AVG","MIN","MAX"}
class ValidationError(Exception):
    pass
def assert_true(cond: bool, msg: str):
    if not cond: raise ValidationError(msg)
def load_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))
def validate_expr(value: Any, allowed_refs: Set[str] | None = None):
    assert_true(isinstance(value, dict), "expression must be object")
    kind = value.get("kind")
    assert_true(kind in {"FUNCTION","FIELD_REF"}, f"unsupported expression kind: {kind}")
    if kind == "FIELD_REF":
        ref, field = value.get("ref"), value.get("field")
        assert_true(isinstance(ref, str) and ref, "FIELD_REF.ref is required")
        assert_true(isinstance(field, str) and field, "FIELD_REF.field is required")
        if allowed_refs is not None: assert_true(ref in allowed_refs, f"FIELD_REF.ref must reference known alias: {ref}")
        return
    name, args = value.get("name"), value.get("args")
    assert_true(isinstance(name, str) and name, "FUNCTION.name is required")
    assert_true(isinstance(args, list), "FUNCTION.args must be array")
    for arg in args:
        if isinstance(arg, dict) and arg.get("kind") in {"FUNCTION","FIELD_REF"}:
            validate_expr(arg, allowed_refs)
def validate_aliases(objects: List[Dict[str, Any]]) -> Set[str]:
    aliases: Set[str] = set()
    for obj in objects:
        assert_true(isinstance(obj, dict), "objects items must be objects")
        assert_true(isinstance(obj.get("objectType"), str) and obj["objectType"], "objectType is required")
        assert_true(isinstance(obj.get("alias"), str) and obj["alias"], "object alias is required")
        assert_true(obj["alias"] not in aliases, f"duplicate object alias: {obj['alias']}")
        aliases.add(obj["alias"])
        if "fromSource" in obj: assert_true(isinstance(obj["fromSource"], str) and obj["fromSource"], "fromSource must be non-empty string")
    return aliases
def validate_relationships(relationships: List[Dict[str, Any]], object_aliases: Set[str]) -> Set[str]:
    rel_aliases: Set[str] = set()
    for rel in relationships:
        assert_true(isinstance(rel, dict), "relationships items must be objects")
        for field in ("relationshipType","alias","from","to"):
            assert_true(isinstance(rel.get(field), str) and rel[field], f"relationship {field} is required")
        assert_true(rel["alias"] not in rel_aliases, f"duplicate relationship alias: {rel['alias']}")
        assert_true(rel["alias"] not in object_aliases, f"relationship alias conflicts with object alias: {rel['alias']}")
        assert_true(rel["from"] in object_aliases, f"relationship.from unknown alias: {rel['from']}")
        assert_true(rel["to"] in object_aliases, f"relationship.to unknown alias: {rel['to']}")
        rel_aliases.add(rel["alias"])
    return rel_aliases
def validate_conditions(node: Dict[str, Any], allowed_refs: Set[str]):
    assert_true(isinstance(node, dict), "conditions must be object")
    kind = node.get("kind")
    assert_true(kind in {"GROUP","PREDICATE"}, "conditions.kind must be GROUP or PREDICATE")
    if kind == "GROUP":
        relation, children = node.get("relation"), node.get("children")
        assert_true(relation in {"AND","OR","NOT"}, "GROUP relation must be AND/OR/NOT")
        assert_true(isinstance(children, list) and children, "GROUP children must be non-empty array")
        if relation == "NOT": assert_true(len(children) == 1, "NOT group must contain exactly one child")
        for child in children: validate_conditions(child, allowed_refs)
        return
    op = node.get("operator")
    assert_true(op in CONDITION_OPERATORS, f"unsupported operator: {op}")
    has_simple = isinstance(node.get("ref"), str) and isinstance(node.get("field"), str)
    has_expr = isinstance(node.get("leftExpr"), dict)
    assert_true(has_simple or has_expr, "PREDICATE requires ref/field or leftExpr")
    if has_simple: assert_true(node["ref"] in allowed_refs, f"PREDICATE ref must be known alias: {node['ref']}")
    if has_expr: validate_expr(node["leftExpr"], allowed_refs)
    expected = CONDITION_OPERATORS[op]
    if expected == 0:
        assert_true("values" not in node, f"operator {op} must not contain values")
    else:
        values = node.get("values")
        assert_true(isinstance(values, list), f"operator {op} requires values array")
        if expected == 1: assert_true(len(values) == 1, f"operator {op} requires exactly one value")
        elif expected == 2: assert_true(len(values) == 2, f"operator {op} requires exactly two values")
        else: assert_true(len(values) >= 1, f"operator {op} requires at least one value")
        for value in values:
            if isinstance(value, dict) and value.get("kind") in {"FUNCTION","FIELD_REF"}: validate_expr(value, allowed_refs)
def validate_returns(items: List[Dict[str, Any]], allowed_refs: Set[str], operation: str):
    assert_true(isinstance(items, list) and items, "returns must be non-empty array")
    metric_count = 0
    for item in items:
        assert_true(isinstance(item, dict), "returns items must be objects")
        kind = item.get("kind")
        assert_true(kind in RETURN_KINDS, f"unsupported returns.kind: {kind}")
        if kind == "FIELDS":
            assert_true(isinstance(item.get("ref"), str) and item["ref"] in allowed_refs, f"FIELDS.ref must be known alias: {item.get('ref')}")
            assert_true(isinstance(item.get("fields"), list) and item["fields"], "FIELDS requires non-empty fields array")
        elif kind == "EXPR":
            validate_expr(item.get("expr"), allowed_refs)
            assert_true(isinstance(item.get("alias"), str) and item["alias"], "EXPR requires alias")
        elif kind == "GROUP_BY":
            assert_true(isinstance(item.get("alias"), str) and item["alias"], "GROUP_BY requires alias")
            if "expr" in item: validate_expr(item["expr"], allowed_refs)
            else:
                assert_true(isinstance(item.get("ref"), str) and item["ref"] in allowed_refs, f"GROUP_BY.ref must be known alias: {item.get('ref')}")
                assert_true(isinstance(item.get("field"), str) and item["field"], "GROUP_BY.field is required")
        elif kind == "METRIC":
            metric_count += 1
            assert_true(isinstance(item.get("ref"), str) and item["ref"] in allowed_refs, f"METRIC.ref must be known alias: {item.get('ref')}")
            assert_true(isinstance(item.get("field"), str) and item["field"], "METRIC.field is required")
            assert_true(item.get("function") in METRIC_FUNCTIONS, f"METRIC.function invalid: {item.get('function')}")
            assert_true(isinstance(item.get("alias"), str) and item["alias"], "METRIC.alias is required")
            if item["function"] != "COUNT": assert_true(item["field"] != "*", "only COUNT may use field='*'")
    if operation == "AGGREGATE":
        assert_true(metric_count >= 1, "AGGREGATE requires at least one METRIC")
        assert_true(all(i.get("kind") in {"GROUP_BY","METRIC"} for i in items), "AGGREGATE returns may contain only GROUP_BY and METRIC")
    if operation == "QUERY":
        assert_true(all(i.get("kind") in {"FIELDS","EXPR"} for i in items), "QUERY returns may contain only FIELDS and EXPR")
def validate_orders(items: List[Dict[str, Any]]):
    assert_true(isinstance(items, list) and items, "orders must be non-empty array")
    for item in items:
        assert_true(isinstance(item, dict), "orders items must be objects")
        assert_true(isinstance(item.get("ref"), str) and item["ref"], "orders.ref is required")
        assert_true(isinstance(item.get("field"), str) and item["field"], "orders.field is required")
        assert_true(item.get("direction") in {"ASC","DESC"}, "orders.direction must be ASC or DESC")
def validate_source_query(items: List[Dict[str, Any]], depth: int):
    assert_true(depth <= 2, "sourceQuery nesting depth exceeds strict limit of 2")
    assert_true(isinstance(items, list) and items, "sourceQuery must be non-empty array")
    seen: Set[str] = set()
    for item in items:
        assert_true(isinstance(item, dict), "sourceQuery items must be objects")
        output_as = item.get("outputAs")
        assert_true(isinstance(output_as, str) and output_as, "sourceQuery.outputAs is required")
        assert_true(output_as not in seen, f"duplicate sourceQuery outputAs: {output_as}")
        seen.add(output_as)
        validate_oql(item, False, depth)
        assert_true(item.get("operation") in READ_OPS, "sourceQuery only supports read operations")
def validate_link_query(block: Dict[str, Any], object_aliases: Set[str]):
    assert_true(isinstance(block, dict), "linkQuery must be object")
    assert_true(block.get("mode") in {"LIST","ONE"}, "linkQuery.mode must be LIST or ONE")
    assert_true(isinstance(block.get("relationshipType"), str) and block["relationshipType"], "linkQuery.relationshipType required")
    for key in ("sourceRef","targetRef"):
        assert_true(isinstance(block.get(key), str) and block[key] in object_aliases, f"linkQuery.{key} must reference known object alias")
    if "direction" in block: assert_true(block["direction"] in {"OUTBOUND","INBOUND","BIDIRECTIONAL"}, "linkQuery.direction invalid")
def validate_mutation(block: Dict[str, Any], operation: str):
    assert_true(isinstance(block, dict), "mutation must be object")
    if operation == "CREATE":
        props = (((block.get("data") or {}).get("properties")) if isinstance(block.get("data"), dict) else None)
        assert_true(isinstance(props, dict) and props, "CREATE requires mutation.data.properties")
    elif operation == "UPDATE":
        assert_true(block.get("scope") in {"ONE","MANY"}, "UPDATE mutation.scope must be ONE or MANY")
        assert_true(isinstance(block.get("set"), dict) and block["set"], "UPDATE requires non-empty mutation.set")
    elif operation == "DELETE":
        assert_true(block.get("scope") in {"ONE","MANY"}, "DELETE mutation.scope must be ONE or MANY")
        assert_true("set" not in block and "data" not in block, "DELETE must not contain mutation.set or mutation.data")
    elif operation == "UPSERT":
        match_by = block.get("matchBy")
        props = (((block.get("data") or {}).get("properties")) if isinstance(block.get("data"), dict) else None)
        assert_true(isinstance(match_by, list) and match_by, "UPSERT requires non-empty mutation.matchBy")
        assert_true(isinstance(props, dict) and props, "UPSERT requires mutation.data.properties")
        for field in match_by: assert_true(field in props, f"UPSERT matchBy field missing from data.properties: {field}")
    elif operation == "BATCH":
        assert_true(isinstance(block.get("atomic"), bool), "BATCH requires boolean mutation.atomic")
        items = block.get("items")
        assert_true(isinstance(items, list) and items, "BATCH requires non-empty mutation.items")
        for item in items:
            assert_true(isinstance(item, dict), "BATCH items must be objects")
            assert_true(item.get("operation") in {"CREATE","UPDATE","DELETE","UPSERT"}, "BATCH items only allow CREATE/UPDATE/DELETE/UPSERT")
            assert_true(item.get("operation") != "BATCH", "nested BATCH not allowed")
            validate_oql(item, True, 0)
def validate_oql(data: Dict[str, Any], is_batch_item: bool = False, depth: int = 0):
    assert_true(isinstance(data, dict), "OQL document must be object")
    unknown = set(data.keys()) - ALLOWED_TOP - {"outputAs"}
    assert_true(not unknown, f"unknown top-level fields: {sorted(unknown)}")
    op = data.get("operation")
    assert_true(op in ALL_OPS, f"operation invalid: {op}")
    if not is_batch_item:
        assert_true(data.get("version") == "1.0", "version must be '1.0'")
        assert_true(isinstance(data.get("schemaRef"), str) and data["schemaRef"], "schemaRef is required")
        if "strict" in data: assert_true(isinstance(data["strict"], bool), "strict must be boolean")
    else:
        for inherited in ("version","schemaRef","strict"): assert_true(inherited not in data, f"BATCH items must not include {inherited}")
    if "maxResults" in data:
        assert_true(isinstance(data["maxResults"], int), "maxResults must be integer")
        assert_true(1 <= data["maxResults"] <= 100000, "maxResults must be between 1 and 100000")
    object_aliases: Set[str] = set(); relationship_aliases: Set[str] = set()
    if op != "BATCH" and "objects" in data:
        assert_true(isinstance(data["objects"], list) and data["objects"], "objects must be non-empty array")
        object_aliases = validate_aliases(data["objects"])
    if op == "ASSOCIATION_QUERY":
        assert_true("relationships" in data, "ASSOCIATION_QUERY requires relationships")
        assert_true(isinstance(data["relationships"], list) and data["relationships"], "relationships must be non-empty array")
        relationship_aliases = validate_relationships(data["relationships"], object_aliases)
    elif "relationships" in data:
        raise ValidationError("relationships only allowed in ASSOCIATION_QUERY")
    if op in READ_OPS: assert_true("returns" in data, f"{op} requires returns")
    if op in {"UPDATE","DELETE"}: assert_true("conditions" in data, f"{op} requires conditions")
    if op in {"CREATE","UPSERT"}: assert_true("conditions" not in data, f"{op} must not include conditions")
    allowed_refs = object_aliases | relationship_aliases
    if "conditions" in data: validate_conditions(data["conditions"], allowed_refs)
    if "returns" in data: validate_returns(data["returns"], allowed_refs, op)
    if "orders" in data: validate_orders(data["orders"])
    if "sourceQuery" in data:
        assert_true(op in READ_OPS, "sourceQuery only allowed in read operations")
        validate_source_query(data["sourceQuery"], depth + 1)
        outputs = {item["outputAs"] for item in data["sourceQuery"]}
        for obj in data.get("objects", []):
            if "fromSource" in obj: assert_true(obj["fromSource"] in outputs, f"objects.fromSource must reference same-level sourceQuery.outputAs: {obj['fromSource']}")
    if op == "LINK_QUERY":
        assert_true(len(data.get("objects", [])) == 2, "LINK_QUERY requires exactly two objects")
        assert_true("linkQuery" in data, "LINK_QUERY requires linkQuery block")
        validate_link_query(data["linkQuery"], object_aliases)
    elif "linkQuery" in data:
        raise ValidationError("linkQuery only allowed in LINK_QUERY")
    if op in WRITE_OPS:
        assert_true("mutation" in data, f"{op} requires mutation")
        validate_mutation(data["mutation"], op)
    elif "mutation" in data:
        raise ValidationError("mutation only allowed in write operations")
    if op in {"CREATE","UPDATE","DELETE","UPSERT"}: assert_true(len(data.get("objects", [])) == 1, f"{op} requires exactly one object")
    if op == "BATCH": assert_true("objects" not in data, "BATCH top level must not contain objects")
def main() -> int:
    parser = argparse.ArgumentParser(description="Validate canonical OQL")
    parser.add_argument("--input", required=True)
    args = parser.parse_args()
    data = load_json(Path(args.input))
    try:
        validate_oql(data)
    except ValidationError as exc:
        print(json.dumps({"success": False, "errors": [{"code": "OQL_VALIDATION_ERROR", "message": str(exc)}]}, ensure_ascii=False, indent=2))
        return 1
    print(json.dumps({"success": True}, ensure_ascii=False, indent=2))
    return 0
if __name__ == "__main__":
    raise SystemExit(main())

File: java/src/main/java/com/onology/oac/queryframework/core/OqlValidator.java
package com.onology.oac.queryframework.core;

import com.onology.oac.queryframework.domain.OqlModels;
import com.onology.oac.queryframework.domain.OqlModels.AggregateFilterKind;
import com.onology.oac.queryframework.domain.OqlModels.ConditionKind;
import com.onology.oac.queryframework.domain.OqlModels.ExpressionKind;
import com.onology.oac.queryframework.domain.OqlModels.OperationType;
import com.onology.oac.queryframework.domain.OqlModels.ReturnKind;
import com.onology.oac.queryframework.domain.ResultModels.OacError;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * Composable OQL validator for query operations.
 */
public class OqlValidator {
    public ValidationResult validate(OqlModels.OqlQuery query) {
        List<OacError> errors = new ArrayList<>();
        if (query == null) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", "query must not be null", "$"));
            return new ValidationResult(errors);
        }
        validateTopLevel(query, errors);
        validateAliasClosure(query, errors);
        validateReturns(query, errors);
        validateCondition(query.conditions(), "conditions", errors);
        validateAggregateFilter(query, errors);
        return new ValidationResult(errors);
    }

    private void validateTopLevel(OqlModels.OqlQuery query, List<OacError> errors) {
        if (query.operation() == null) {
            errors.add(OacError.of("UNKNOWN_OPERATION", "operation must not be null", "operation"));
            return;
        }
        if (query.schemaRef() == null || query.schemaRef().isBlank()) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", "schemaRef must not be blank", "schemaRef"));
        }
        if ((query.operation() == OperationType.QUERY || query.operation() == OperationType.AGGREGATE)
                && (query.objects() == null || query.objects().isEmpty())) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", query.operation() + " requires objects", "objects"));
        }
        if ((query.operation() == OperationType.QUERY || query.operation() == OperationType.AGGREGATE
                || query.operation() == OperationType.ASSOCIATION_QUERY)
                && (query.returns() == null || query.returns().isEmpty())) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", query.operation() + " requires returns", "returns"));
        }
        if (query.operation() == OperationType.QUERY && query.relationships() != null && !query.relationships().isEmpty()) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", "QUERY must not contain relationships", "relationships"));
        }
        if (query.operation() == OperationType.QUERY && query.aggregateFilter() != null) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", "QUERY must not contain aggregateFilter", "aggregateFilter"));
        }
        if (query.operation() == OperationType.AGGREGATE && query.relationships() != null && !query.relationships().isEmpty()) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", "AGGREGATE must not contain relationships", "relationships"));
        }
        if (query.operation() == OperationType.ASSOCIATION_QUERY
                && (query.relationships() == null || query.relationships().isEmpty())) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", "ASSOCIATION_QUERY requires relationships", "relationships"));
        }
        if (query.operation() == OperationType.ASSOCIATION_QUERY && query.aggregateFilter() != null) {
            errors.add(OacError.of("OQL_VALIDATION_ERROR", "ASSOCIATION_QUERY must not contain aggregateFilter", "aggregateFilter"));
        }
    }

    private void validateAliasClosure(OqlModels.OqlQuery query, List<OacError> errors) {
        Set<String> aliases = new HashSet<>();
        if (query.objects() != null) {
            for (int i = 0; i < query.objects().size(); i++) {
                OqlModels.OqlObject object = query.objects().get(i);
                if (object.alias() == null || object.alias().isBlank()) {
                    errors.add(OacError.of("UNKNOWN_ALIAS", "object alias must not be blank", "objects[" + i + "].alias"));
                } else if (!aliases.add(object.alias())) {
                    errors.add(OacError.of("UNKNOWN_ALIAS", "duplicate alias: " + object.alias(), "objects[" + i + "].alias"));
                }
            }
        }
        if (query.relationships() != null) {
            for (int i = 0; i < query.relationships().size(); i++) {
                OqlModels.OqlRelationship relationship = query.relationships().get(i);
                if (relationship.alias() != null && aliases.contains(relationship.alias())) {
                    errors.add(OacError.of("UNKNOWN_ALIAS", "relationship alias conflicts with object alias: "
                            + relationship.alias(), "relationships[" + i + "].alias"));
                }
                if (!aliases.contains(relationship.from())) {
                    errors.add(OacError.of("UNKNOWN_ALIAS", "relationships.from references unknown alias: "
                            + relationship.from(), "relationships[" + i + "].from"));
                }
                if (!aliases.contains(relationship.to())) {
                    errors.add(OacError.of("UNKNOWN_ALIAS", "relationships.to references unknown alias: "
                            + relationship.to(), "relationships[" + i + "].to"));
                }
            }
        }
        validateConditionAliases(query.conditions(), aliases, "conditions", errors);
        if (query.returns() != null) {
            for (int i = 0; i < query.returns().size(); i++) {
                OqlModels.OqlReturnItem item = query.returns().get(i);
                if ((item.kind() == ReturnKind.FIELDS || item.kind() == ReturnKind.METRIC || item.kind() == ReturnKind.GROUP_BY)
                        && item.ref() != null && !aliases.contains(item.ref())) {
                    errors.add(OacError.of("UNKNOWN_ALIAS", "returns.ref references unknown alias: " + item.ref(),
                            "returns[" + i + "].ref"));
                }
            }
        }
    }

    private void validateReturns(OqlModels.OqlQuery query, List<OacError> errors) {
        if (query.returns() == null) {
            return;
        }
        boolean hasMetric = false;
        for (int i = 0; i < query.returns().size(); i++) {
            OqlModels.OqlReturnItem item = query.returns().get(i);
            if (item.kind() == null) {
                errors.add(OacError.of("INVALID_RETURN_KIND", "returns.kind must not be null", "returns[" + i + "].kind"));
                continue;
            }
            if (query.operation() == OperationType.AGGREGATE) {
                if (item.kind() != ReturnKind.GROUP_BY && item.kind() != ReturnKind.METRIC) {
                    errors.add(OacError.of("INVALID_RETURN_KIND", "AGGREGATE only allows GROUP_BY and METRIC",
                            "returns[" + i + "].kind"));
                }
                if (item.kind() == ReturnKind.METRIC) {
                    hasMetric = true;
                    if (item.function() == null) {
                        errors.add(OacError.of("INVALID_RETURN_KIND", "METRIC requires function", "returns[" + i + "].function"));
                    }
                    if (item.function() != OqlModels.AggregateFunction.COUNT && Objects.equals(item.field(), "*")) {
                        errors.add(OacError.of("INVALID_RETURN_KIND", "only COUNT allows field=*", "returns[" + i + "].field"));
                    }
                }
                if ((item.kind() == ReturnKind.METRIC || item.kind() == ReturnKind.GROUP_BY)
                        && (item.alias() == null || item.alias().isBlank())) {
                    errors.add(OacError.of("INVALID_RETURN_KIND", item.kind() + " requires alias", "returns[" + i + "].alias"));
                }
            } else if (item.kind() != ReturnKind.FIELDS && item.kind() != ReturnKind.EXPR) {
                errors.add(OacError.of("INVALID_RETURN_KIND", query.operation() + " only allows FIELDS and EXPR",
                        "returns[" + i + "].kind"));
            }
            if (item.kind() == ReturnKind.FIELDS && (item.fields() == null || item.fields().isEmpty())) {
                errors.add(OacError.of("INVALID_RETURN_KIND", "FIELDS requires explicit fields", "returns[" + i + "].fields"));
            }
            if ((item.kind() == ReturnKind.EXPR || item.kind() == ReturnKind.GROUP_BY) && item.expr() != null) {
                validateExpression(item.expr(), "returns[" + i + "].expr", errors);
            }
        }
        if (query.operation() == OperationType.AGGREGATE && !hasMetric) {
            errors.add(OacError.of("INVALID_RETURN_KIND", "AGGREGATE requires at least one METRIC", "returns"));
        }
    }

    private void validateCondition(OqlModels.OqlCondition condition, String path, List<OacError> errors) {
        if (condition == null) {
            return;
        }
        if (condition.kind() == ConditionKind.GROUP) {
            if (condition.children() == null || condition.children().isEmpty()) {
                errors.add(OacError.of("INVALID_CONDITION", "GROUP.children must not be empty", path + ".children"));
            } else {
                if (condition.relation() == OqlModels.GroupRelation.NOT && condition.children().size() != 1) {
                    errors.add(OacError.of("INVALID_CONDITION", "NOT group requires exactly one child", path + ".children"));
                }
                for (int i = 0; i < condition.children().size(); i++) {
                    validateCondition(condition.children().get(i), path + ".children[" + i + "]", errors);
                }
            }
        } else if (condition.kind() == ConditionKind.PREDICATE) {
            validateOperatorValues(condition.operator(), condition.values(), path, errors, "INVALID_CONDITION");
            if (condition.left() != null) {
                validateExpression(condition.left(), path + ".left", errors);
            }
        }
    }

    private void validateAggregateFilter(OqlModels.OqlQuery query, List<OacError> errors) {
        if (query.aggregateFilter() == null) {
            return;
        }
        if (query.operation() != OperationType.AGGREGATE) {
            errors.add(OacError.of("INVALID_AGGREGATE_FILTER", "aggregateFilter only allowed for AGGREGATE", "aggregateFilter"));
            return;
        }
        Set<String> metricAliases = new HashSet<>();
        if (query.returns() != null) {
            query.returns().stream()
                    .filter(r -> r.kind() == ReturnKind.METRIC)
                    .map(OqlModels.OqlReturnItem::alias)
                    .filter(Objects::nonNull)
                    .forEach(metricAliases::add);
        }
        validateAggregateFilterNode(query.aggregateFilter(), "aggregateFilter", metricAliases, errors);
    }

    private void validateAggregateFilterNode(OqlModels.OqlAggregateFilter filter, String path, Set<String> metricAliases,
                                             List<OacError> errors) {
        if (filter.kind() == AggregateFilterKind.METRIC_PREDICATE) {
            if (!metricAliases.contains(filter.metricAlias())) {
                errors.add(OacError.of("INVALID_AGGREGATE_FILTER", "metricAlias not found: " + filter.metricAlias(),
                        path + ".metricAlias"));
            }
            validateOperatorValues(filter.operator(), filter.values(), path, errors, "INVALID_AGGREGATE_FILTER");
        } else if (filter.kind() == AggregateFilterKind.GROUP) {
            if (filter.children() == null || filter.children().isEmpty()) {
                errors.add(OacError.of("INVALID_AGGREGATE_FILTER", "GROUP.children must not be empty", path + ".children"));
            } else {
                if (filter.relation() == OqlModels.GroupRelation.NOT && filter.children().size() != 1) {
                    errors.add(OacError.of("INVALID_AGGREGATE_FILTER", "NOT group requires exactly one child", path + ".children"));
                }
                for (int i = 0; i < filter.children().size(); i++) {
                    validateAggregateFilterNode(filter.children().get(i), path + ".children[" + i + "]", metricAliases, errors);
                }
            }
        }
    }

    private void validateConditionAliases(OqlModels.OqlCondition condition, Set<String> aliases, String path,
                                          List<OacError> errors) {
        if (condition == null) {
            return;
        }
        if (condition.ref() != null && !aliases.contains(condition.ref())) {
            errors.add(OacError.of("UNKNOWN_ALIAS", "condition ref references unknown alias: " + condition.ref(), path + ".ref"));
        }
        if (condition.left() != null) {
            validateExpressionAlias(condition.left(), aliases, path + ".left", errors);
        }
        if (condition.children() != null) {
            for (int i = 0; i < condition.children().size(); i++) {
                validateConditionAliases(condition.children().get(i), aliases, path + ".children[" + i + "]", errors);
            }
        }
    }

    private void validateExpressionAlias(OqlModels.OqlExpression expression, Set<String> aliases, String path,
                                         List<OacError> errors) {
        if (expression.kind() == ExpressionKind.FIELD && expression.ref() != null && !aliases.contains(expression.ref())) {
            errors.add(OacError.of("UNKNOWN_ALIAS", "FIELD.ref references unknown alias: " + expression.ref(), path + ".ref"));
        }
        if (expression.args() != null) {
            for (int i = 0; i < expression.args().size(); i++) {
                validateExpressionAlias(expression.args().get(i), aliases, path + ".args[" + i + "]", errors);
            }
        }
    }

    private void validateExpression(OqlModels.OqlExpression expression, String path, List<OacError> errors) {
        if (expression == null || expression.kind() == null) {
            errors.add(OacError.of("FUNCTION_NOT_SUPPORTED", "expression.kind must not be null", path));
            return;
        }
        if (expression.kind() == ExpressionKind.FUNCTION) {
            if (expression.name() == null || expression.name().isBlank()) {
                errors.add(OacError.of("UNREGISTERED_FUNCTION", "function name must not be blank", path + ".name"));
            }
            if (expression.name() != null && Set.of("COUNT", "SUM", "AVG", "MIN", "MAX").contains(expression.name())) {
                errors.add(OacError.of("FUNCTION_NOT_SUPPORTED", "aggregate function must use returns.kind=METRIC", path));
            }
        }
        if (expression.args() != null) {
            for (int i = 0; i < expression.args().size(); i++) {
                validateExpression(expression.args().get(i), path + ".args[" + i + "]", errors);
            }
        }
    }

    private void validateOperatorValues(OqlModels.Operator operator, List<Object> values, String path, List<OacError> errors,
                                        String code) {
        if (operator == null) {
            errors.add(OacError.of(code, "operator must not be null", path + ".operator"));
            return;
        }
        if (operator == OqlModels.Operator.BETWEEN && (values == null || values.size() != 2)) {
            errors.add(OacError.of(code, "BETWEEN requires exactly two values", path + ".values"));
        }
        if (Set.of(OqlModels.Operator.IS_NULL, OqlModels.Operator.IS_NOT_NULL, OqlModels.Operator.IS_EMPTY,
                OqlModels.Operator.IS_NOT_EMPTY, OqlModels.Operator.EXISTS, OqlModels.Operator.NOT_EXISTS).contains(operator)
                && values != null && !values.isEmpty()) {
            errors.add(OacError.of(code, operator + " must not contain values", path + ".values"));
        }
    }

    public record ValidationResult(List<OacError> errors) {
        public boolean isSuccess() {
            return errors == null || errors.isEmpty();
        }
    }
}
